README.md
Compare System Versions
=======================

Check package and frameworks versions within projects running in production.

This library look up on each project and scan for `composer.json` and `package.json` 
files to check if the versions of its installed  packages are outdated.

[screenshot]: ./screenshot.png "Screenshot"

![Screenshot][screenshot]

## Setup

Create a fresh composer project then install this project.

    mkdir /var/www/compare-system-versions
    composer init
    
    git init
    echo "/vendor/" > .gitignore
    git add .gitignore
    git commit -m "First Commit" 

## Install

Download the latest release `csv.phar`.

    curl -Lo csv https://github.com/adrianorsouza/compare-system-versions/releases/download/v0.1.0/csv.phar
    chmod +x csv
    sudo mv csv /usr/local/bin/

## Config

First Create a config YAML file `config.yaml`:

```yaml
system_versions:
    latest:
      php: 7.4.7
      nginx: 1.18.0
      mysql: 8.0.20
      platform: Ubuntu 20.04 LTS (Focal Fossa)

    current:
      php: 7.3.1
      nginx: 1.16.0
      mysql: 8.0.20
      platform: Ubuntu 18.04 LTS (Bionic Beaver)

package_versions:
    latest:
      phpunit: 9.2.0
      laravel: 7.18.0
      react: 16.13.1
      grunt: 1.2.1

iterator:
  dir: /var/www
  packages:
    - composer.lock
    - package-lock.json
    - bower.json
  exclude:
    - bower_components
    - node_modules
    - public
    - public_html
    - vendor
    - compare-system-versions

options:
  title: System Versions Compare
  build_dir: public/
```

## Usage

    php csv.phar update config.yaml
    
NOTES ABOUT CONFIG:

- `build_dir` are always relative to the current directory, so don't define it as absolute path.

The code above will scan for projects within `/var/www` dir and get the versions for packages and versions like: 

- PHP
- Ubuntu
- Laravel Framework
- ReactJS Framework
- GruntJS Task Runner


## Author

**Adriano Rosa** (https://adrianorosa.com)  

## Licence

Copyright © 2020, Adriano Rosa  <info@adrianorosa.com>
All rights reserved.

For the full copyright and license information, please view the LICENSE 
file that was distributed within the source root of this project.


CHANGELOG.md
# Changelog

Version 0.1.0 - 2020-07-23
===========================

  - 2020-07-23 - add npm release script
  - 2020-07-23 - include the HTML frontend when building the store data
  - 2020-07-23 - Update box.json
  - 2020-07-23 - Update .gitignore
  - 2020-07-23 - Create CHANGELOG.md
  - 2020-07-23 - Update README.md
  - 2020-07-23 - add updated_at to the store data
  - 2020-07-22 - Update README.md


Version 0.1.0 - 2020-07-23
==========================

   - First Release 


src/Parser/Parser.php
<?php

namespace Adrianorosa\Csv\Parser;

use Adrianorosa\Csv\Version;

/**
 * Class Parser.
 *
 * @author Adriano Rosa <https://adrianorosa.com>
 * @date   2020-07-04 23:45
 *
 * @package Adrianorosa\Csv\Parser
 */
class Parser
{

    protected $config;

    /**
     * @var array
     */
    protected $data = [];

    /**
     * @var array
     */
    protected $result = [];

    /**
     * Parser constructor.
     *
     * @param  array $data
     * @param  \Adrianorosa\Csv\Parser\Config $config
     */
    public function __construct(array $data, Config $config)
    {
        $this->data = $data;
        $this->config = $config;
    }

    /**
     *
     */
    public function parse()
    {
        foreach ($this->data as $path => $items) {
            $result = [
                'name' => $path,
                // 'path' => $items['dirname'],
                'files' => [
                    // 'composer' => null,
                    // 'package' => null,
                    // 'bower' => null,
                ],
            ];

            foreach ($items as $key => $value) {
                if ($key === 'dirname') {
                    continue;
                }

                $content = json_decode(file_get_contents($value), true);
                switch ($key) {
                    case 'composer':
                        $result['files']['composer'] = $this->getComposerPackagesVersion($content);
                        break;

                    case 'package':
                        if ($gruntVersion = $this->getNPMPackagesVersion($content, 'grunt')) {
                            $result['files']['package']['grunt'] = [
                                'version' => $gruntVersion,
                                'satisfies' => Version::compare($this->config->getPackageVersionsLatest('grunt'), $gruntVersion)
                            ];
                        };
                        if ($reactVersion = $this->getNPMPackagesVersion($content, 'react')) {
                            $result['files']['package']['react'] = [
                                'version' => $reactVersion,
                                'satisfies' => Version::compare($this->config->getPackageVersionsLatest('react'), $reactVersion),
                            ];
                        };
                        break;

                    case 'bower':
                        if ($jquery = $this->getBowerPackageVersion($content, 'jquery')) {
                            $result['files']['bower']['jquery']['version'] = $jquery;
                        };
                        if ($bootstrap = $this->getBowerPackageVersion($content, 'bootstrap-sass')) {
                            $result['files']['bower']['bootstrap-sass']['version'] = $bootstrap;
                        };
                        break;
                }
            }

            $this->result[] = $result;
        }

        return $this;
    }

    /**
     * @param  array $content
     *
     * @return array|null
     */
    protected function getComposerPackagesVersion(array $content): ?array
    {
        $fn = function ($p, $v) {
            if ($p['name'] === 'phpunit/phpunit') {
                return [
                    'version' => $v,
                    'satisfies' => Version::compare($this->config->getPackageVersionsLatest('phpunit'), $v)
                ];
            }
            return null;
        };

        $composer = [];
        // Look for dependencies within packages list
        foreach ($content['packages'] as $package) {
            $v = $this->formatVersion($package['version']);
            if ($package['name'] === 'laravel/framework') {
                $composer['laravel'] = [
                    'version' => $v,
                    'satisfies' => Version::compare($this->config->getPackageVersionsLatest('laravel'), $v)
                ];
                break;
            }

            if ($phpunit = $fn($package, $v)) {
                $composer['phpunit'] = $phpunit;
                break;
            }
        }

        // Look for dependencies within dev packages list
        foreach ($content['packages-dev'] as $packages_dev) {
            $v = $this->formatVersion($packages_dev['version']);
            if ($phpunit = $fn($packages_dev, $v)) {
                $composer['phpunit'] = $phpunit;
                break;
            }
        }

        return $composer;
    }

    /**
     * @param array $content
     * @param $packageName
     *
     * @return string|null
     */
    protected function getNPMPackagesVersion(array $content, $packageName): ?string
    {
        $version = null;
        foreach ($content['dependencies'] as $package => $value) {
            if ($package === $packageName) {
                $version = $this->formatVersion($value['version']);
                break;
            }
        }

        return $version;
    }

    /**
     * @param array $content
     * @param $packageName
     *
     * @return string|null
     */
    protected function getBowerPackageVersion(array $content, $packageName): ?string
    {
        $version = null;
        foreach ($content['devDependencies'] as $package => $installedVersion) {
            if ($package === $packageName) {
                $version = $this->formatVersion($installedVersion);
                break;
            }
        }

        return $version;
    }

    /**
     * @param $version
     *
     * @return string|string[]
     */
    protected function formatVersion($version)
    {
        return str_replace(['v', '~', '^', ' ', '>='], '', $version);
    }

    /**
     * @return array
     */
    public function getScannedProjects()
    {
        return $this->result;
    }
}
